# Project 2 - Data Cleanup and Preparation
## Rebecca Lewis
### DSC 680

In [1]:
#import libraries
import pandas as pd
import numpy as np
import datetime


In [2]:
alldata = pd.read_excel('../Data/tags_delivered.xlsx', sheet_name=None)
alldata.keys()

C:\Users\rvick\Anaconda3\envs\R_Python\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Your version of xlrd is 1.2.0. In xlrd >= 2.0, only the xls format is supported. As a result, the openpyxl engine will be used if it is installed and the engine argument is not specified. Install openpyxl instead.
  """Entry point for launching an IPython kernel.


dict_keys(['2015', '2016', '2017', '2018', '2019', '2020', '2021', 'Contact Info', 'Sheet1', 'Sheet2', 'ESRI_MAPINFO_SHEET'])

In [3]:
sheet_names=['2015', '2016', '2017', '2018', '2019', '2020']

tag_data = pd.concat([pd.read_excel('../Data/tags_delivered.xlsx', sheet_name=s).assign(sheet_name=s) for s in sheet_names])
tag_data.head()

C:\Users\rvick\Anaconda3\envs\R_Python\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Your version of xlrd is 1.2.0. In xlrd >= 2.0, only the xls format is supported. As a result, the openpyxl engine will be used if it is installed and the engine argument is not specified. Install openpyxl instead.
  This is separate from the ipykernel package so we can avoid doing imports until


Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
0        NaN           500001        501000  2014-12-12 00:00:00  1000.0   
1        NaN           501001        501200  2014-12-13 00:00:00   200.0   
2        NaN           501201        501300  2014-12-29 00:00:00   100.0   
3        NaN           501201        501283  2016-01-05 00:00:00   -83.0   
4        NaN           501301        501400  2014-12-24 00:00:00   100.0   

   Returned                 Clinic Name Unnamed: 7 Unnamed: 8 sheet_name  ...  \
0       NaN  Southern Animal Foundation        NaN        NaN       2015  ...   
1       NaN     Banfield 1023 (Elmwood)        NaN        NaN       2015  ...   
2       NaN          Animal Care Center        NaN        NaN       2015  ...   
3  Returned          Animal Care Center        NaN        NaN       2015  ...   
4       NaN       Allstar Animal Clinic        NaN        NaN       2015  ...   

  Unnamed: 10  Unnamed: 11 Unnamed: 12 Unnamed: 13  Unnamed: 14  Unnamed: 15  \
0         NaN          NaN         NaN         NaN          NaN          NaN   
1         NaN          NaN         NaN         NaN          NaN          NaN   
2         NaN          NaN         NaN         NaN          NaN          NaN   
3         NaN          NaN         NaN         NaN          NaN          NaN   
4         NaN          NaN         NaN         NaN          NaN          NaN   

   Unnamed: 16  Unnamed: 17  Unnamed: 18 Unnamed: 6  
0          NaN          NaN          NaN        NaN  
1          NaN          NaN          NaN        NaN  
2          NaN          NaN          NaN        NaN  
3          NaN          NaN          NaN        NaN  
4          NaN          NaN          NaN        NaN  

[5 rows x 21 columns]

In [4]:
tag_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1541 entries, 0 to 283
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        4 non-null      object 
 1   Beginning tags #  1534 non-null   object 
 2   Ending tag #      1532 non-null   object 
 3   Delivered         1524 non-null   object 
 4   Volume            1537 non-null   float64
 5   Returned          547 non-null    object 
 6   Clinic Name       1529 non-null   object 
 7   Unnamed: 7        10 non-null     object 
 8   Unnamed: 8        28 non-null     object 
 9   sheet_name        1541 non-null   object 
 10  Unnamed: 9        4 non-null      object 
 11  Unnamed: 10       8 non-null      float64
 12  Unnamed: 11       4 non-null      object 
 13  Unnamed: 12       7 non-null      object 
 14  Unnamed: 13       0 non-null      float64
 15  Unnamed: 14       0 non-null      float64
 16  Unnamed: 15       0 non-null      float64
 

In [5]:
#Drop columns with 0 non-null values

tag_data = tag_data.drop(columns=['Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15', 'Unnamed: 17'], axis=1)
tag_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1541 entries, 0 to 283
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        4 non-null      object 
 1   Beginning tags #  1534 non-null   object 
 2   Ending tag #      1532 non-null   object 
 3   Delivered         1524 non-null   object 
 4   Volume            1537 non-null   float64
 5   Returned          547 non-null    object 
 6   Clinic Name       1529 non-null   object 
 7   Unnamed: 7        10 non-null     object 
 8   Unnamed: 8        28 non-null     object 
 9   sheet_name        1541 non-null   object 
 10  Unnamed: 9        4 non-null      object 
 11  Unnamed: 10       8 non-null      float64
 12  Unnamed: 11       4 non-null      object 
 13  Unnamed: 12       7 non-null      object 
 14  Unnamed: 16       1 non-null      float64
 15  Unnamed: 18       1 non-null      object 
 16  Unnamed: 6        2 non-null      object 
d

In [6]:
#view unnamed columns with non null values
tag_data[tag_data['Unnamed: 6'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
194        NaN           854501        854570  2019-01-14 00:00:00    70.0   
82         NaN           021201        021300  2019-12-30 00:00:00   100.0   

    Returned          Clinic Name Unnamed: 7 Unnamed: 8 sheet_name Unnamed: 9  \
194      NaN  NOLA Animal Clinic         NaN        NaN       2018        NaT   
82       NaN   Animal Care Center        NaN        NaN       2020        NaT   

     Unnamed: 10 Unnamed: 11 Unnamed: 12  Unnamed: 16 Unnamed: 18 Unnamed: 6  
194          NaN         NaN         NaN          NaN         NaN             
82           NaN         NaN         NaN          NaN         NaN     Closed

In [7]:
#irrelevant to the analysis; can drop this column
tag_data = tag_data.drop(columns=['Unnamed: 6'], axis=1)

In [8]:
tag_data[tag_data['Unnamed: 18'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
167        NaN           715269        715287  2017-12-27 00:00:00   -19.0   

     Returned          Clinic Name Unnamed: 7 Unnamed: 8 sheet_name  \
167  Returned  Marrero Vet. Clinic        NaN        NaN       2017   

    Unnamed: 9  Unnamed: 10 Unnamed: 11 Unnamed: 12  Unnamed: 16  \
167        NaN          NaN         NaN         NaN        500.0   

                     Unnamed: 18  
167  Louisiana SPCA Wellness TBI

In [9]:
#doesn't make sense; dropping the column
tag_data = tag_data.drop(columns=['Unnamed: 18'], axis=1)

In [10]:
tag_data[tag_data['Unnamed: 16'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
167        NaN           715269        715287  2017-12-27 00:00:00   -19.0   

     Returned          Clinic Name Unnamed: 7 Unnamed: 8 sheet_name  \
167  Returned  Marrero Vet. Clinic        NaN        NaN       2017   

    Unnamed: 9  Unnamed: 10 Unnamed: 11 Unnamed: 12  Unnamed: 16  
167        NaN          NaN         NaN         NaN        500.0

In [11]:
#referring to the same record above
tag_data = tag_data.drop(columns=['Unnamed: 16'], axis=1)

In [12]:
tag_data[tag_data['Unnamed: 12'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
174        NaN           853318        853500  2019-12-23 00:00:00  -183.0   
175        NaN           853501        854000  2019-01-23 00:00:00  -500.0   
176        NaN           840XXX        840900  2019-01-23 00:00:00   -15.0   
177        NaN           806001        806500  2017-12-26 00:00:00   500.0   
178        NaN           821501        822000  2018-02-22 00:00:00   500.0   
179        NaN           826701        827000  2018-05-04 00:00:00   300.0   
180        NaN           842001        842500  2018-07-18 00:00:00   500.0   

     Returned                     Clinic Name Unnamed: 7 Unnamed: 8  \
174  Returned  Metairie Small Animal Hospital   808001.0   809000.0   
175  Returned  Metairie Small Animal Hospital   836001.0   836500.0   
176  Returned  Metairie Small Animal Hospital   842501.0   843000.0   
177       NaN     Midcity Veterinary Hospital   842801.0   843000.0   
178       NaN     Midcity Veterinary Hospital        NaN        NaN   
179       NaN     Midcity Veterinary Hospital        NaN        NaN   
180       NaN     Midcity Veterinary Hospital        NaN        NaN   

    sheet_name           Unnamed: 9  Unnamed: 10 Unnamed: 11      Unnamed: 12  
174       2018  2017-12-28 00:00:00       1000.0         NaN  Pet Care Center  
175       2018  2018-05-29 00:00:00        500.0         NaN  Pet Care Center  
176       2018  2018-07-18 00:00:00        500.0         NaN  Pet Care Center  
177       2018  2019-02-25 00:00:00       -200.0    Returned  Pet Care Center  
178       2018                  NaT        -32.0    Returned  Pet Care Center  
179       2018                  NaT        -40.0    Returned  Pet Care Center  
180       2018                  NaT        -41.0    Returned  Pet Care Center

The data in these columns appear to be "scratch" work performed in the margins to get calculations as needed.  The tag numbers shown are represented elsewehre in the data.

In [13]:
#dropping this column
tag_data = tag_data.drop(columns=['Unnamed: 12'], axis=1)

In [14]:
tag_data[tag_data['Unnamed: 11'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
177        NaN           806001        806500  2017-12-26 00:00:00   500.0   
178        NaN           821501        822000  2018-02-22 00:00:00   500.0   
179        NaN           826701        827000  2018-05-04 00:00:00   300.0   
180        NaN           842001        842500  2018-07-18 00:00:00   500.0   

    Returned                  Clinic Name Unnamed: 7 Unnamed: 8 sheet_name  \
177      NaN  Midcity Veterinary Hospital   842801.0   843000.0       2018   
178      NaN  Midcity Veterinary Hospital        NaN        NaN       2018   
179      NaN  Midcity Veterinary Hospital        NaN        NaN       2018   
180      NaN  Midcity Veterinary Hospital        NaN        NaN       2018   

              Unnamed: 9  Unnamed: 10 Unnamed: 11  
177  2019-02-25 00:00:00       -200.0    Returned  
178                  NaT        -32.0    Returned  
179                  NaT        -40.0    Returned  
180                  NaT        -41.0    Returned

In [15]:
#same observations as above
tag_data = tag_data.drop(columns=['Unnamed: 11'], axis=1)

In [16]:
tag_data[tag_data['Unnamed: 10'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
174        NaN           853318        853500  2019-12-23 00:00:00  -183.0   
175        NaN           853501        854000  2019-01-23 00:00:00  -500.0   
176        NaN           840XXX        840900  2019-01-23 00:00:00   -15.0   
177        NaN           806001        806500  2017-12-26 00:00:00   500.0   
178        NaN           821501        822000  2018-02-22 00:00:00   500.0   
179        NaN           826701        827000  2018-05-04 00:00:00   300.0   
180        NaN           842001        842500  2018-07-18 00:00:00   500.0   
181        NaN           850001        850500  2018-09-26 00:00:00   500.0   

     Returned                     Clinic Name Unnamed: 7 Unnamed: 8  \
174  Returned  Metairie Small Animal Hospital   808001.0   809000.0   
175  Returned  Metairie Small Animal Hospital   836001.0   836500.0   
176  Returned  Metairie Small Animal Hospital   842501.0   843000.0   
177       NaN     Midcity Veterinary Hospital   842801.0   843000.0   
178       NaN     Midcity Veterinary Hospital        NaN        NaN   
179       NaN     Midcity Veterinary Hospital        NaN        NaN   
180       NaN     Midcity Veterinary Hospital        NaN        NaN   
181       NaN     Midcity Veterinary Hospital        NaN        NaN   

    sheet_name           Unnamed: 9  Unnamed: 10  
174       2018  2017-12-28 00:00:00       1000.0  
175       2018  2018-05-29 00:00:00        500.0  
176       2018  2018-07-18 00:00:00        500.0  
177       2018  2019-02-25 00:00:00       -200.0  
178       2018                  NaT        -32.0  
179       2018                  NaT        -40.0  
180       2018                  NaT        -41.0  
181       2018                  NaT       1687.0

In [17]:
#appears to be scratch work noted above
tag_data = tag_data.drop(columns=['Unnamed: 10'], axis=1)

In [18]:
tag_data[tag_data['Unnamed: 9'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
174        NaN           853318        853500  2019-12-23 00:00:00  -183.0   
175        NaN           853501        854000  2019-01-23 00:00:00  -500.0   
176        NaN           840XXX        840900  2019-01-23 00:00:00   -15.0   
177        NaN           806001        806500  2017-12-26 00:00:00   500.0   

     Returned                     Clinic Name Unnamed: 7 Unnamed: 8  \
174  Returned  Metairie Small Animal Hospital   808001.0   809000.0   
175  Returned  Metairie Small Animal Hospital   836001.0   836500.0   
176  Returned  Metairie Small Animal Hospital   842501.0   843000.0   
177       NaN     Midcity Veterinary Hospital   842801.0   843000.0   

    sheet_name           Unnamed: 9  
174       2018  2017-12-28 00:00:00  
175       2018  2018-05-29 00:00:00  
176       2018  2018-07-18 00:00:00  
177       2018  2019-02-25 00:00:00

In [19]:
#same observations as above
tag_data = tag_data.drop(columns=['Unnamed: 9'], axis=1)

In [20]:
tag_data[tag_data['Unnamed: 8'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
6          NaN           501501        502500  2014-12-18 00:00:00  1000.0   
7          NaN           501753        501800  2015-01-13 00:00:00   -48.0   
131        NaN           518501        518600  2015-04-24 00:00:00   100.0   
135        NaN           519501        518564  2016-03-04 00:00:00    64.0   
156        NaN           527001        528000  2015-05-04 00:00:00  1000.0   
206        NaN           541001        541500  2015-09-29 00:00:00   500.0   
207        NaN           541401        541500  2016-01-05 00:00:00  -100.0   
209        NaN           542001        542500  2015-09-30 00:00:00   500.0   
174        NaN           853318        853500  2019-12-23 00:00:00  -183.0   
175        NaN           853501        854000  2019-01-23 00:00:00  -500.0   
176        NaN           840XXX        840900  2019-01-23 00:00:00   -15.0   
177        NaN           806001        806500  2017-12-26 00:00:00   500.0   
184        NaN           815068        815091  2018-12-26 00:00:00   -24.0   
185        NaN           815113        815200  2018-12-26 00:00:00   -88.0   
186        NaN           817701        817800  2018-01-08 00:00:00   100.0   
187        NaN           832201        832300  2018-05-17 00:00:00   100.0   
188        NaN           852901        853000  2018-11-16 00:00:00   100.0   
216        NaN           953401        953600  2019-10-02 00:00:00   200.0   
217        NaN           953501        953600  2019-12-30 00:00:00  -100.0   
218        NaN           941501        942000  2019-07-03 00:00:00   500.0   
219        NaN           941901        942000  2019-12-30 00:00:00  -100.0   
220        NaN           916901        917200  2018-12-27 00:00:00   300.0   
221        NaN           917135        917200  2019-12-30 00:00:00   -66.0   
222        NaN           918301        918500  2018-12-26 00:00:00   200.0   
223        NaN           918425        918426  2020-01-09 00:00:00    -2.0   
224        NaN           918427        918500  2019-12-30 00:00:00   -74.0   
225        NaN           912501        913000  2018-12-26 00:00:00   500.0   
95         NaN           022101        022200  2020-01-09 00:00:00   100.0   

     Returned                           Clinic Name Unnamed: 7 Unnamed: 8  \
6         NaN                 Metairie Small Animal        NaN      -35.0   
7    Returned                 Metairie Small Animal        NaN      -76.0   
131       NaN             Marrero Veterinary Clinic        NaN      500.0   
135  Returned                               Marrero        NaN     1000.0   
156       NaN                 Metairie Small Animal        NaN     -100.0   
206       NaN             Maple Small Animal Clinic        NaN     1000.0   
207  Returned             Maple Small Animal Clinic        NaN      -48.0   
209       NaN                 Metairie Small Animal        NaN      952.0   
174  Returned        Metairie Small Animal Hospital   808001.0   809000.0   
175  Returned        Metairie Small Animal Hospital   836001.0   836500.0   
176  Returned        Metairie Small Animal Hospital   842501.0   843000.0   
177       NaN           Midcity Veterinary Hospital   842801.0   843000.0   
184  Returned                      Natural Pet Care        NaN     -200.0   
185  Returned                      Natural Pet Care        NaN      -32.0   
186       NaN             New Orleans Animal Health        NaN      -40.0   
187       NaN             New Orleans Animal Health        NaN      -41.0   
188       NaN             New Orleans Animal Health        NaN     -313.0   
216       NaN  Thrive Affordable Vet Care (Midcity)        NaN      -35.0   
217  Returned  Thrive Affordable Vet Care (Midcity)        NaN      -66.0   
218       NaN     Thrive Affordable Vet Care (Vets)        NaN      -92.0   
219  Returned     Thrive Affordable Vet Care (Vets)        NaN      -74.0   
220       NaN            Uptown Veterinary Hosp

In [21]:
#these values don't make sense with the line, any tag numbers are represented elsewhere
tag_data = tag_data.drop(columns=['Unnamed: 8'], axis=1)

In [22]:
tag_data[tag_data['Unnamed: 7'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
161        NaN           530001        530500  2015-05-29 00:00:00   500.0   
37         NaN           748909        748973  2018-01-03 00:00:00   -64.0   
107        NaN           716161        716175                  NaN   -25.0   
134        NaN           742119        742191  2017-12-27 00:00:00    73.0   
143        NaN           716901        716914  2017-12-27 00:00:00   -14.0   
253        NaN           727213        727225                  NaN   -12.0   
174        NaN           853318        853500  2019-12-23 00:00:00  -183.0   
175        NaN           853501        854000  2019-01-23 00:00:00  -500.0   
176        NaN           840XXX        840900  2019-01-23 00:00:00   -15.0   
177        NaN           806001        806500  2017-12-26 00:00:00   500.0   

     Returned                               Clinic Name       Unnamed: 7  \
161       NaN                 Crescent City Veterinary                     
37   Returned  Banifield 1034 (West Bank) Special issue       per shelly   
107  Returned                    Gretna Animal Hospital       Shelly saw   
134  Returned                  Louisiana SPCA Adoptions  rt by Dr. Smith   
143  Returned               Louisiana SPCA Wellness TBI  rt by Dr. Smith   
253  Returned                                       NaN        Not found   
174  Returned            Metairie Small Animal Hospital         808001.0   
175  Returned            Metairie Small Animal Hospital         836001.0   
176  Returned            Metairie Small Animal Hospital         842501.0   
177       NaN               Midcity Veterinary Hospital         842801.0   

    sheet_name  
161       2015  
37        2017  
107       2017  
134       2017  
143       2017  
253       2017  
174       2018  
175       2018  
176       2018  
177       2018

In [23]:
#includes random notes and tag numbers that are represented elsewhere
tag_data = tag_data.drop(columns=['Unnamed: 7'], axis=1)

In [24]:
tag_data[tag_data['Unnamed: 0'].isna()==False]

Unnamed: 0 Beginning tags # Ending tag #             Delivered  Volume  \
154          X           707001        707500  2016-12-26 00:00:00   500.0   
192          X           701001        701500  2016-12-22 00:00:00   500.0   
206          X           700001        701000  2016-12-26 00:00:00  1000.0   
235          X           707501        708000  2016-12-28 00:00:00   500.0   

    Returned                    Clinic Name sheet_name  
154      NaN  Magazine Street Animal Clinic       2017  
192      NaN                Pet Care Center       2017  
206      NaN     Southern Animal Foundation       2017  
235      NaN   VCA Causeway Animal Hospital       2017

In [25]:
#can drop 
tag_data = tag_data.drop(columns=['Unnamed: 0'], axis=1)

### Before I can start to summarize, I need to investigate the categorical data and standardize.

In [26]:
tag_data.describe(include=np.object).T

count unique                  top freq
Beginning tags #  1534   1439               513801    5
Ending tag #      1532   1286               516800    5
Delivered         1524    582  2021-01-04 00:00:00  114
Returned           547      6             Returned  539
Clinic Name       1529    157     Community Clinic   67
sheet_name        1541      6                 2020  284

### the most tags were distributed in 2020
### The community clinic has the most tag deliveries and returns
### There are 157 clinics noted, need to view and standardize
### there are 6 unique values in the Returned column - this should only be 1 'Returned'.  will need to standardized.
### the dataset has 1544 records but there are only 1532 and 1534 with tag numbers.  May be some blank columns.  Also only 1524 in the date column


In [27]:
tag_data.describe().T

count        mean         std    min   25%    50%    75%     max
Volume  1537.0  193.735849  371.871429 -500.0 -24.0  100.0  500.0  9639.0

### only 1538 records with tag volume
### the max value for volumn is 9639 which leads me to believe there are summary values in the spreadsheet being imported as observations


In [28]:
#update column names
tag_data.columns

Index(['Beginning tags #', 'Ending tag # ', 'Delivered', 'Volume', 'Returned',
       'Clinic Name', 'sheet_name'],
      dtype='object')

In [29]:
tag_data.columns = ['beginning_tag_num', 'ending_tag_num', 'delivered', 'volume', 'returned',
       'clinic', 'tag_year']
tag_data.head()

beginning_tag_num ending_tag_num            delivered  volume  returned  \
0            500001         501000  2014-12-12 00:00:00  1000.0       NaN   
1            501001         501200  2014-12-13 00:00:00   200.0       NaN   
2            501201         501300  2014-12-29 00:00:00   100.0       NaN   
3            501201         501283  2016-01-05 00:00:00   -83.0  Returned   
4            501301         501400  2014-12-24 00:00:00   100.0       NaN   

                       clinic tag_year  
0  Southern Animal Foundation     2015  
1     Banfield 1023 (Elmwood)     2015  
2          Animal Care Center     2015  
3          Animal Care Center     2015  
4       Allstar Animal Clinic     2015

In [30]:
#view observations missing beginning tag_num, ending_tag_num, volume, and delivered
tag_data[(tag_data['beginning_tag_num'].isna()) | (tag_data['ending_tag_num'].isna()) | (tag_data['volume'].isna()) | (tag_data['delivered'].isna())]

beginning_tag_num ending_tag_num            delivered  volume  returned  \
42             507322            NaN  2015-11-15 00:00:00    -1.0  Returned   
54             510067         510100                  NaN   -34.0  Returned   
168            635385         635500                  NaN  -116.0  Returned   
181            639801          64000                  NaN  -200.0  Returned   
107            716161         716175                  NaN   -25.0  Returned   
245            719324         7193xx                  NaN   -48.0  Returned   
246            719509         7195xx                  NaN  -100.0  Returned   
247            719804         7198xx                  NaN  -100.0  Returned   
249            752901         753000                  NaN  -100.0  Returned   
253            727213         727225                  NaN   -12.0  Returned   
201            842701            NaN  2019-05-14 00:00:00   -25.0   Rturned   
203               NaN            NaN                  NaN   -32.0  Returned   
204               NaN            NaN                  NaN   -40.0  Returned   
205               NaN            NaN                  NaN   -41.0  Returned   
5              002201         002300                  NaT     NaN       NaN   
280               NaN            NaN                  NaT     NaN       NaN   
281               NaN            NaN                  NaT     NaN       NaN   
282               NaN            NaN                  NaT     NaN       NaN   
283               NaN            NaN                  NaT  9639.0       NaN   

                           clinic tag_year  
42       Home Veterinary Services     2015  
54        W. Esplanade Vet Clinic     2015  
168  VCA Causeway Animal Hospital     2016  
181  VCA Causeway Animal Hospital     2016  
107        Gretna Animal Hospital     2017  
245                 Vetco @ Petco     2017  
246                 Vetco @ Petco     2017  
247                 Vetco @ Petco     2017  
249                 Vetco @ Petco     2017  
253                           NaN     2017  
201               Pet Care Center     2018  
203               Pet Care Center     2018  
204               Pet Care Center     2018  
205               Pet Care Center     2018  
5                             NaN     2020  
280                           NaN     2020  
281                           NaN     2020  
282                           NaN     2020  
283                           NaN     2020

In [31]:
#if only 1 of the tag numbers is missing we can keep.  I'll revise the condition
tag_data[((tag_data['beginning_tag_num'].isna()) & (tag_data['ending_tag_num'].isna())) | (tag_data['volume'].isna()) | (tag_data['delivered'].isna())]

beginning_tag_num ending_tag_num delivered  volume  returned  \
54             510067         510100       NaN   -34.0  Returned   
168            635385         635500       NaN  -116.0  Returned   
181            639801          64000       NaN  -200.0  Returned   
107            716161         716175       NaN   -25.0  Returned   
245            719324         7193xx       NaN   -48.0  Returned   
246            719509         7195xx       NaN  -100.0  Returned   
247            719804         7198xx       NaN  -100.0  Returned   
249            752901         753000       NaN  -100.0  Returned   
253            727213         727225       NaN   -12.0  Returned   
203               NaN            NaN       NaN   -32.0  Returned   
204               NaN            NaN       NaN   -40.0  Returned   
205               NaN            NaN       NaN   -41.0  Returned   
5              002201         002300       NaT     NaN       NaN   
280               NaN            NaN       NaT     NaN       NaN   
281               NaN            NaN       NaT     NaN       NaN   
282               NaN            NaN       NaT     NaN       NaN   
283               NaN            NaN       NaT  9639.0       NaN   

                           clinic tag_year  
54        W. Esplanade Vet Clinic     2015  
168  VCA Causeway Animal Hospital     2016  
181  VCA Causeway Animal Hospital     2016  
107        Gretna Animal Hospital     2017  
245                 Vetco @ Petco     2017  
246                 Vetco @ Petco     2017  
247                 Vetco @ Petco     2017  
249                 Vetco @ Petco     2017  
253                           NaN     2017  
203               Pet Care Center     2018  
204               Pet Care Center     2018  
205               Pet Care Center     2018  
5                             NaN     2020  
280                           NaN     2020  
281                           NaN     2020  
282                           NaN     2020  
283                           NaN     2020

In [32]:
#i'm going to allow delivered to be null because it seems there are valid volumes for those. 
tag_data[((tag_data['beginning_tag_num'].isna()) & (tag_data['ending_tag_num'].isna())) | (tag_data['volume'].isna())]

beginning_tag_num ending_tag_num delivered  volume  returned  \
203               NaN            NaN       NaN   -32.0  Returned   
204               NaN            NaN       NaN   -40.0  Returned   
205               NaN            NaN       NaN   -41.0  Returned   
5              002201         002300       NaT     NaN       NaN   
280               NaN            NaN       NaT     NaN       NaN   
281               NaN            NaN       NaT     NaN       NaN   
282               NaN            NaN       NaT     NaN       NaN   
283               NaN            NaN       NaT  9639.0       NaN   

              clinic tag_year  
203  Pet Care Center     2018  
204  Pet Care Center     2018  
205  Pet Care Center     2018  
5                NaN     2020  
280              NaN     2020  
281              NaN     2020  
282              NaN     2020  
283              NaN     2020

In [33]:
#if there is a clinic listed, they are still valid
tag_data[(((tag_data['beginning_tag_num'].isna()) & (tag_data['ending_tag_num'].isna())) | (tag_data['volume'].isna())) & (tag_data['clinic'].isna())]

beginning_tag_num ending_tag_num delivered  volume returned clinic  \
5              002201         002300       NaT     NaN      NaN    NaN   
280               NaN            NaN       NaT     NaN      NaN    NaN   
281               NaN            NaN       NaT     NaN      NaN    NaN   
282               NaN            NaN       NaT     NaN      NaN    NaN   
283               NaN            NaN       NaT  9639.0      NaN    NaN   

    tag_year  
5       2020  
280     2020  
281     2020  
282     2020  
283     2020

In [34]:
#all of these observations can be removed.  They have values with no context for the clinic.  Some appear to be blank lines, some appear to be totals, some appear to be inclomplete data

drop_indexes = tag_data[(((tag_data['beginning_tag_num'].isna()) & (tag_data['ending_tag_num'].isna())) | (tag_data['volume'].isna())) & (tag_data['clinic'].isna())].index

In [35]:
drop_indexes

Int64Index([5, 280, 281, 282, 283], dtype='int64')

In [36]:
tag_data = tag_data.drop(drop_indexes).reset_index(drop=True)


In [37]:
tag_data.shape

(1531, 7)

In [38]:
tag_data.describe()

volume
count  1531.000000
mean    187.154148
std     283.401228
min    -500.000000
25%     -24.000000
50%     100.000000
75%     500.000000
max    2500.000000

In [39]:
# i want to view the max volume records
tag_data[tag_data['volume']==2500]

beginning_tag_num ending_tag_num            delivered  volume returned  \
237            590001         592500  2015-01-15 00:00:00  2500.0      NaN   

                               clinic tag_year  
237  Community Clinic (Out of Parish)     2015

In [40]:
#view distinct return values
tag_data.returned.unique()

array([nan, 'Returned', 'Destroyed',
       '                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           

In [41]:
#update Rturned to returned
tag_data['returned'] = tag_data['returned'].replace('Rturned', 'Returned')

In [42]:
#view distinct return values
tag_data.returned.unique()

array([nan, 'Returned', 'Destroyed',
       '                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           

In [43]:
#view Destroyed
tag_data[tag_data['returned']=='Destroyed']

beginning_tag_num ending_tag_num            delivered  volume   returned  \
11              503731         503750  2015-03-10 00:00:00    20.0  Destroyed   
1249            928966         929000  2019-04-25 00:00:00   -35.0  Destroyed   

         clinic tag_year  
11    Destroyed     2015  
1249        NaN     2019

In [44]:
#they do not have clinics associated with them and there are only two - removing from the dataset
drop_indexes = tag_data[tag_data['returned']=='Destroyed'].index
tag_data = tag_data.drop(drop_indexes).reset_index(drop=True)

In [45]:
tag_data.returned.unique()

array([nan, 'Returned',
       '                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

In [46]:
#view voided
tag_data[tag_data['returned']=='VOID']

beginning_tag_num ending_tag_num            delivered  volume returned  \
484            649701         650000  2016-12-30 00:00:00   300.0     VOID   
486            690501         690550                 VOID    50.0     VOID   
938            825341         825342  2018-04-09 00:00:00    -2.0     VOID   

                         clinic tag_year  
484                         NaN     2016  
486  Benson Veterinary Care LLC     2016  
938                  Per Shelly     2018

In [47]:
# The ones not associated to a clinic can be removed.
# The one associated to a clinic were not issued to a pet so I will count them as returned.
drop_indexes = [484, 938]
tag_data = tag_data.drop(drop_indexes).reset_index(drop=True)


In [48]:
tag_data.returned.unique()

array([nan, 'Returned',
       '                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

In [49]:
#view returned-misc
tag_data[tag_data['returned']=='Returned-Misc']

beginning_tag_num ending_tag_num            delivered  volume  \
243            5365XX         5366XX  2015-01-11 00:00:00   -28.0   

          returned         clinic tag_year  
243  Returned-Misc  Vetco @ Petco     2015

In [50]:
#replacing with Returned
tag_data['returned'] = tag_data['returned'].replace('Returned-Misc', 'Returned')

In [51]:
tag_data.returned.unique()

array([nan, 'Returned',
       '                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

In [52]:
#see if i can strip the spaces out of the one that takes up multiple lines
tag_data['returned'] = tag_data['returned'].str.strip()

In [53]:
tag_data.returned.unique()

array([nan, 'Returned', '', 'VOID'], dtype=object)

In [54]:
#view the nan and '' values
tag_data[(tag_data['returned']=='') | (tag_data['returned'].isna())]

beginning_tag_num ending_tag_num            delivered  volume returned  \
0               500001         501000  2014-12-12 00:00:00  1000.0      NaN   
1               501001         501200  2014-12-13 00:00:00   200.0      NaN   
2               501201         501300  2014-12-29 00:00:00   100.0      NaN   
4               501301         501400  2014-12-24 00:00:00   100.0      NaN   
5               501501         502500  2014-12-18 00:00:00  1000.0      NaN   
...                ...            ...                  ...     ...      ...   
1499            056701         057000  2021-11-19 00:00:00   300.0      NaN   
1503            057012         057100  2021-01-04 00:00:00   100.0      NaN   
1504            057101         057200  2020-12-14 00:00:00   100.0      NaN   
1505            057201         057500  2020-12-08 00:00:00   300.0      NaN   
1509            057601         057700  2020-12-28 00:00:00   200.0      NaN   

                                 clinic tag_year  
0            Southern Animal Foundation     2015  
1               Banfield 1023 (Elmwood)     2015  
2                    Animal Care Center     2015  
4                 Allstar Animal Clinic     2015  
5                 Metairie Small Animal     2015  
...                                 ...      ...  
1499  Crescent City Veterinary Hospital     2020  
1503              Allstar Animal Clinic     2020  
1504         Southern Animal Foundation     2020  
1505        Midcity Veterinary Hospital     2020  
1509                 NOLA Animal Clinic     2020  

[985 rows x 7 columns]

In [55]:
#create new column for action: either delivered returned or voided
tag_data['action'] = tag_data['returned']
tag_data.loc[tag_data['returned'].isna(), 'action'] = 'Delivered'

In [56]:
tag_data.action.unique()

array(['Delivered', 'Returned', '', 'VOID'], dtype=object)

In [57]:
#view ''
tag_data[tag_data.action == '']

beginning_tag_num ending_tag_num            delivered  volume returned  \
97            516501         516800  2015-03-28 00:00:00   300.0            

                 clinic tag_year action  
97  E12 (Franklin Ave.)     2015

In [58]:
tag_data['action'] = tag_data['action'].replace('', 'Delivered')

In [59]:
#see if any delivered actions have associated negative volumes
tag_data[(tag_data['volume'] < 0) & (tag_data['action']=='Delivered') ]

beginning_tag_num ending_tag_num            delivered  volume returned  \
769            845501         845684  2018-12-26 00:00:00  -185.0      NaN   
777            844401         844458  2018-12-26 00:00:00   -58.0      NaN   
897            851801         852000  2018-12-26 00:00:00  -200.0      NaN   
979            817632         817634  2018-12-26 00:00:00    -3.0      NaN   

                          clinic tag_year     action  
769      Banfield 1023 (Elmwood)     2018  Delivered  
777     Banfield 1175 (Veterans)     2018  Delivered  
897    Maple Small Animal Clinic     2018  Delivered  
979  VCA Airline Animal Hospital     2018  Delivered

In [60]:
#Update them to returned
tag_data.loc[(tag_data['volume'] < 0) & (tag_data['action']=='Delivered'),'action']= 'Returned'

#view to make sure it worked
tag_data[(tag_data['volume'] < 0) & (tag_data['action']=='Delivered') ]

Empty DataFrame
Columns: [beginning_tag_num, ending_tag_num, delivered, volume, returned, clinic, tag_year, action]
Index: []

In [61]:
#find returned values that have positive volume
tag_data[(tag_data['volume'] > 0) & (tag_data['action']=='Returned') ]

beginning_tag_num ending_tag_num            delivered  volume  returned  \
30              505401         505454  2016-03-23 00:00:00    54.0  Returned   
133             519501         518564  2016-03-04 00:00:00    64.0  Returned   
353             620901         620967  2016-03-23 00:00:00    67.0  Returned   
486             612445         612445  2017-12-20 00:00:00     1.0  Returned   
487             612447         612450  2017-12-20 00:00:00     4.0  Returned   
590             705642         705700  2018-02-22 00:00:00    52.0  Returned   
601             715974         716000  2018-01-03 00:00:00    27.0  Returned   
621             742119         742191  2017-12-27 00:00:00    73.0  Returned   
664             751412         751500  2017-12-27 00:00:00    89.0  Returned   
1144            917687         917700  2020-01-03 00:00:00    14.0  Returned   

                              clinic tag_year    action  
30        Convenient Home Veterinary     2015  Returned  
133                          Marrero     2015  Returned  
353                E24 (Poland Ave.)     2016  Returned  
486      Terry Parkway Animal Clinic     2016  Returned  
487      Terry Parkway Animal Clinic     2016  Returned  
590          Fur De Lis Vet Services     2017  Returned  
601         Home Veterinary Services     2017  Returned  
621         Louisiana SPCA Adoptions     2017  Returned  
664   Metairie Small Animal Hospital     2017  Returned  
1144             Marrero Vet. Clinic     2019  Returned

In [62]:
#because the user deliberately typed the returned value in the returned field. these records will have the volume updated to be negative
tag_data.loc[(tag_data['volume'] > 0) & (tag_data['action']=='Returned'),'volume']= -1*tag_data['volume']


In [63]:
#see if it worked
tag_data[(tag_data['volume'] > 0) & (tag_data['action']=='Returned') ]

Empty DataFrame
Columns: [beginning_tag_num, ending_tag_num, delivered, volume, returned, clinic, tag_year, action]
Index: []

In [64]:
#see if there are any missing values
clinic_summary = tag_data.groupby(['clinic']).count().reset_index()

In [65]:
clinic_summary.clinic.unique()

array(['?????????????????', '????????????????????????',
       'Acadiana West Animal Clinic', 'Algiers Animal Clinic',
       'All Cats Mobile Clinic', 'Allstar Animal Clinic',
       'Animal Care Center', 'Animal Medical & Surgical Hospital',
       'Ark Animal Hospital', 'At Home - Bourgeois',
       'At Home Veterinary Care', 'Aurora Animal Hospital',
       'Avenue Animal Wellness and Emergency',
       'Avian & Exotic Animal Hospital of Louisiana', 'Banfield - Remote',
       'Banfield 1023 (Elmwood)', 'Banfield 1034 (West Bank)',
       'Banfield 1034 (West Bank) ',
       'Banfield 1034 (West Bank) Special Issue',
       'Banfield 1034 (Westbank)', 'Banfield 1175 (Veterans)',
       'Banifield 1034 (West Bank) Special issue',
       'Batt Veterinary Services', 'Benson Veterinary Care LLC',
       'Big Easy Animal Rescue', 'Canal Street Animal Clinic',
       'Canal Street Veterinary ', 'Canal Street Veterinary Hospital',
       'Carrollton Veterinary', 'Carrollton Veterinary Hos

In [66]:
#set unknown clinics to unknown
tag_data['clinic'] = tag_data['clinic'].replace(['?????????????????', '????????????????????????', 'XYZ','XYZ Space Holder'], 'Unknown')


In [67]:
#rename duplicates and misnomers
tag_data['clinic'] = tag_data['clinic'].replace('XYZ Space Holder (Villalobos)','Villalobos')
tag_data['clinic'] = tag_data['clinic'].replace(['Banfield 1034 (West Bank) ','Banfield 1034 (West Bank) Special Issue',
                                                 'Banfield 1034 (Westbank)','Banifield 1034 (West Bank) Special issue'],'Banfield 1034 (West Bank)')
tag_data['clinic'] = tag_data['clinic'].replace('Canal Street Veterinary ', 'Canal Street Veterinary Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Carrollton Veterinary', 'Carrollton Veterinary Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Chalmette Pet Wellness', 'Chalmette Pet Wellness Clinic & Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Clearview Veterinary', 'Clearview Veterinary Hospital') 
tag_data['clinic'] = tag_data['clinic'].replace('Convenient Home Veterinary', 'Convenient Home Veterinary Care')
tag_data['clinic'] = tag_data['clinic'].replace(['Crescent City Veterinary ','Crescent City Veterinary Hospital Special Issue'],
                                                'Crescent City Veterinary Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Dog Day Afternoon','Dog Day Afternoon, LLC')
tag_data['clinic'] = tag_data['clinic'].replace('Doskey Mobile Veterinary', 'Doskey Mobile Veterinary Care')
tag_data['clinic'] = tag_data['clinic'].replace('Fur de Lis', 'Fur De Lis Vet Services')
tag_data['clinic'] = tag_data['clinic'].replace('Hickory Knoll Veterinary','Hickory Knoll Veterinary Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('house calls!', 'House Calls!')
tag_data['clinic'] = tag_data['clinic'].replace('Lakeside Animal Hostital', 'Lakeside Animal Hospital')
tag_data['clinic'] = tag_data['clinic'].replace(['Magazine Street Aniaml Clinic','Magazine Street Animal'], 'Magazine Street Animal Clinic')
tag_data['clinic'] = tag_data['clinic'].replace(['Marrero', 'Marrero Vet. Clinic'], 'Marrero Veterinary Clinic')
tag_data['clinic'] = tag_data['clinic'].replace('Midcity Veterinary Hospital ', 'Midcity Veterinary Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Metairie Small Animal', 'Metairie Small Animal Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('NOLA Animal Clinic ', 'NOLA Animal Clinic')
tag_data['clinic'] = tag_data['clinic'].replace('Pet Care Center - J.P.', 'Pet Care Center')
tag_data['clinic'] = tag_data['clinic'].replace('NOPD Dwyer', 'E0 NOPD Dwyer')
tag_data['clinic'] = tag_data['clinic'].replace('Uptown Vet Hospital', 'Uptown Veterinary Hospital')
tag_data['clinic'] = tag_data['clinic'].replace(['VCA Airline Animal','VCA Airline Animal '], 'VCA Airline Animal Hospital')
tag_data['clinic'] = tag_data['clinic'].replace(['VCA Causway','VCA Causeway Animal'], 'VCA Causeway Animal Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Vetco @ Vetco  ','Vetco @ Petco')  
tag_data['clinic'] = tag_data['clinic'].replace('Westside Animal Hospital','Westside Veterinary Hospital') 
tag_data['clinic'] = tag_data['clinic'].replace(['LA-SPCA (Special Issue)', 'Louisiana SPCA (Adoptions)', 
                                                 'Louisiana SPCA (Client Care)', 'Louisiana SPCA (TBI)', 
                                                 'Louisiana SPCA (TBI) (client care?)', 'Louisiana SPCA (Wellness TBI)', 
                                                 'Louisiana SPCA Adoptions', 'Louisiana SPCA Client Care', 
                                                 'Louisiana SPCA Wellness TBI', 'Wellness TBI', 'Community Clinic',
                                                 'Community Clinic (Out of Parish)','Community Clinic (Special Issue)'], 'Louisiana SPCA')                 

In [68]:
tag_data['clinic'] = tag_data['clinic'].replace('At Home - Bourgeois', 'At Home Veterinary Care')



In [69]:
tag_data['clinic'].value_counts()

Louisiana SPCA                    157
Metairie Small Animal Hospital     56
Magazine Street Animal Clinic      44
Midcity Veterinary Hospital        42
Prytania Veterinary Hospital       41
                                 ... 
Canal Street Animal Clinic          1
Geaux Nola Pets Mobile              1
E24 (Poland Ave)                    1
Primary Veterinary Care             1
All Cats Mobile Clinic              1
Name: clinic, Length: 104, dtype: int64

In [70]:
#investigate others
tag_data[tag_data['clinic'] == 'Thrive Affordable Vet Care (Midcity)']

beginning_tag_num ending_tag_num            delivered  volume  returned  \
1218            953401         953600  2019-10-02 00:00:00   200.0       NaN   
1219            953501         953600  2019-12-30 00:00:00  -100.0  Returned   
1291            016501         017000  2019-12-26 00:00:00   500.0       NaN   
1389            031501         032000  2020-05-13 00:00:00   500.0       NaN   
1453            050501         051000  2020-09-09 00:00:00   500.0       NaN   
1454            050639         050700  2021-01-04 00:00:00   -62.0  Returned   
1455            050701         051000  2021-01-04 00:00:00  -300.0  Returned   

                                    clinic tag_year     action  
1218  Thrive Affordable Vet Care (Midcity)     2019  Delivered  
1219  Thrive Affordable Vet Care (Midcity)     2019   Returned  
1291  Thrive Affordable Vet Care (Midcity)     2020  Delivered  
1389  Thrive Affordable Vet Care (Midcity)     2020  Delivered  
1453  Thrive Affordable Vet Care (Midcity)     2020  Delivered  
1454  Thrive Affordable Vet Care (Midcity)     2020   Returned  
1455  Thrive Affordable Vet Care (Midcity)     2020   Returned

In [71]:
tag_data[tag_data['clinic'] == 'Thrive Affordable Vet Care']

beginning_tag_num ending_tag_num            delivered  volume returned  \
1217            924001         924100  2019-03-12 00:00:00   100.0      NaN   

                          clinic tag_year     action  
1217  Thrive Affordable Vet Care     2019  Delivered

In [72]:
#based on activity for that year, the thrive without a location will be assigned to midcity
tag_data['clinic'] = tag_data['clinic'].replace('Thrive Affordable Vet Care', 'Thrive Affordable Vet Care (Midcity)')

In [73]:
# these appear to be part of the annual rabies drive.  They will be lumped into one category since
# we are not focused on the performance of those locations
# 'E0 (Louisiana SPCA', 'E0 NOPD Dwyer', 'E1 (Magazine St)',
#        'E1 (Magazine St.)', 'E12 (Franklin Ave)', 'E12 (Franklin Ave.)',
#        'E16 (MLK)', 'E17 (Woodland)', 'E24 (Poland Ave)',
#        'E24 (Poland Ave.)', 'E25 (S. Carrollton)', 'E27 (Elysian Fields)',
#        'E35 (N. Carrollton)', 'E36 (Read Blvd)', 'E36 (Read Blvd.)'
  
tag_data['clinic'] = tag_data['clinic'].replace(['E0 (Louisiana SPCA', 'E0 NOPD Dwyer', 'E1 (Magazine St)',
       'E1 (Magazine St.)', 'E12 (Franklin Ave)', 'E12 (Franklin Ave.)',
       'E16 (MLK)', 'E17 (Woodland)', 'E24 (Poland Ave)',
       'E24 (Poland Ave.)', 'E25 (S. Carrollton)', 'E27 (Elysian Fields)',
       'E35 (N. Carrollton)', 'E36 (Read Blvd)', 'E36 (Read Blvd.)'], 'Annual Rabies Vaccination Drive')
   


In [74]:
#investigate random values

tag_data[tag_data['clinic'] == 'Spares Lori']


beginning_tag_num ending_tag_num            delivered  volume  returned  \
960            826501         826700  2018-03-26 00:00:00  -200.0  Returned   
961            826501         826700  2018-03-23 00:00:00   200.0       NaN   

          clinic tag_year     action  
960  Spares Lori     2018   Returned  
961  Spares Lori     2018  Delivered

In [75]:
#these can be eliminated
tag_data = tag_data[tag_data['clinic'] != 'Spares Lori']

In [76]:
tag_data[tag_data['clinic'] == 'Miscellaneous']

beginning_tag_num ending_tag_num            delivered  volume  returned  \
1163            955601         955700  2020-01-14 00:00:00  -100.0  Returned   

             clinic tag_year    action  
1163  Miscellaneous     2019  Returned

In [77]:
#set as unknown 
tag_data['clinic'] = tag_data['clinic'].replace('Miscellaneous', 'Unknown')

In [78]:
#try to convert delivered to action_year, action_month
tag_data['action_date'] = tag_data['delivered'].apply(pd.to_datetime, format='%Y-%d-%m %H:%M')

ValueError: time data '12/272014' does not match format '%Y-%d-%m %H:%M' (match)

In [79]:
#errors - 12/272014
tag_data[tag_data['delivered'] == '12/272014']

beginning_tag_num ending_tag_num  delivered  volume returned  \
21            504501         504600  12/272014   100.0      NaN   

                          clinic tag_year     action  
21  Manhattan Animal Clinic, LLC     2015  Delivered

In [80]:
tag_data['delivered'] = tag_data['delivered'].replace('12/272014', '2014-12-27 00:00:00')

In [81]:
tag_data['action_date'] = tag_data['delivered'].apply(pd.to_datetime, infer_datetime_format=True, errors='coerce')

In [82]:
tag_data.head()

beginning_tag_num ending_tag_num            delivered  volume  returned  \
0            500001         501000  2014-12-12 00:00:00  1000.0       NaN   
1            501001         501200  2014-12-13 00:00:00   200.0       NaN   
2            501201         501300  2014-12-29 00:00:00   100.0       NaN   
3            501201         501283  2016-01-05 00:00:00   -83.0  Returned   
4            501301         501400  2014-12-24 00:00:00   100.0       NaN   

                       clinic tag_year     action action_date  
0  Southern Animal Foundation     2015  Delivered  2014-12-12  
1     Banfield 1023 (Elmwood)     2015  Delivered  2014-12-13  
2          Animal Care Center     2015  Delivered  2014-12-29  
3          Animal Care Center     2015   Returned  2016-01-05  
4       Allstar Animal Clinic     2015  Delivered  2014-12-24

In [83]:
#find null action_dates
tag_data[tag_data['action_date'].isna()]['delivered'].unique()

array([nan, '1/6//2015', datetime.datetime(3314, 11, 17, 0, 0), 'VOID',
       '12/272017', '12/272018', '1/52019'], dtype=object)

In [84]:
#correct dates
tag_data['delivered'] = tag_data['delivered'].replace('1/6//2015','2015-01-06')
tag_data['delivered'] = tag_data['delivered'].replace('3314-11-17 00:00:00','2014-11-17 00:00:00')
tag_data['delivered'] = tag_data['delivered'].replace('12/272017','2017-12-27')
tag_data['delivered'] = tag_data['delivered'].replace('12/272018','2018-12-27')
tag_data['delivered'] = tag_data['delivered'].replace('1/52019', '2019-01-05')



In [85]:
#reconvert delivered column
tag_data['action_date'] = tag_data['delivered'].apply(pd.to_datetime, infer_datetime_format=True, errors='coerce')

In [86]:
tag_data[tag_data['action_date'].isna()]

beginning_tag_num ending_tag_num            delivered  volume  returned  \
52             510067         510100                  NaN   -34.0  Returned   
100            516701         516800  3314-11-17 00:00:00    75.0       NaN   
412            635385         635500                  NaN  -116.0  Returned   
425            639801          64000                  NaN  -200.0  Returned   
485            690501         690550                 VOID    50.0      VOID   
594            716161         716175                  NaN   -25.0  Returned   
732            719324         7193xx                  NaN   -48.0  Returned   
733            719509         7195xx                  NaN  -100.0  Returned   
734            719804         7198xx                  NaN  -100.0  Returned   
736            752901         753000                  NaN  -100.0  Returned   
740            727213         727225                  NaN   -12.0  Returned   
942               NaN            NaN                  NaN   -32.0  Returned   
943               NaN            NaN                  NaN   -40.0  Returned   
944               NaN            NaN                  NaN   -41.0  Returned   

                           clinic tag_year     action action_date  
52        W. Esplanade Vet Clinic     2015   Returned         NaT  
100           The SULA Foundation     2015  Delivered         NaT  
412  VCA Causeway Animal Hospital     2016   Returned         NaT  
425  VCA Causeway Animal Hospital     2016   Returned         NaT  
485    Benson Veterinary Care LLC     2016       VOID         NaT  
594        Gretna Animal Hospital     2017   Returned         NaT  
732                 Vetco @ Petco     2017   Returned         NaT  
733                 Vetco @ Petco     2017   Returned         NaT  
734                 Vetco @ Petco     2017   Returned         NaT  
736                 Vetco @ Petco     2017   Returned         NaT  
740                           NaN     2017   Returned         NaT  
942               Pet Care Center     2018   Returned         NaT  
943               Pet Care Center     2018   Returned         NaT  
944               Pet Care Center     2018   Returned         NaT

In [87]:
tag_data['delivered'] = tag_data['delivered'].replace(datetime.datetime(3314, 11, 17, 0, 0), '2014-11-17 00:00:00')

In [88]:
tag_data['action_date'] = tag_data['delivered'].apply(pd.to_datetime, infer_datetime_format=True, errors='coerce')

In [89]:
tag_data[tag_data['action_date'].isna()]

beginning_tag_num ending_tag_num delivered  volume  returned  \
52             510067         510100       NaN   -34.0  Returned   
412            635385         635500       NaN  -116.0  Returned   
425            639801          64000       NaN  -200.0  Returned   
485            690501         690550      VOID    50.0      VOID   
594            716161         716175       NaN   -25.0  Returned   
732            719324         7193xx       NaN   -48.0  Returned   
733            719509         7195xx       NaN  -100.0  Returned   
734            719804         7198xx       NaN  -100.0  Returned   
736            752901         753000       NaN  -100.0  Returned   
740            727213         727225       NaN   -12.0  Returned   
942               NaN            NaN       NaN   -32.0  Returned   
943               NaN            NaN       NaN   -40.0  Returned   
944               NaN            NaN       NaN   -41.0  Returned   

                           clinic tag_year    action action_date  
52        W. Esplanade Vet Clinic     2015  Returned         NaT  
412  VCA Causeway Animal Hospital     2016  Returned         NaT  
425  VCA Causeway Animal Hospital     2016  Returned         NaT  
485    Benson Veterinary Care LLC     2016      VOID         NaT  
594        Gretna Animal Hospital     2017  Returned         NaT  
732                 Vetco @ Petco     2017  Returned         NaT  
733                 Vetco @ Petco     2017  Returned         NaT  
734                 Vetco @ Petco     2017  Returned         NaT  
736                 Vetco @ Petco     2017  Returned         NaT  
740                           NaN     2017  Returned         NaT  
942               Pet Care Center     2018  Returned         NaT  
943               Pet Care Center     2018  Returned         NaT  
944               Pet Care Center     2018  Returned         NaT

In [90]:
tag_data['action_year'] = tag_data['action_date'].dt.year
tag_data['action_month'] = tag_data['action_date'].dt.month

In [91]:
tag_data.head()

beginning_tag_num ending_tag_num            delivered  volume  returned  \
0            500001         501000  2014-12-12 00:00:00  1000.0       NaN   
1            501001         501200  2014-12-13 00:00:00   200.0       NaN   
2            501201         501300  2014-12-29 00:00:00   100.0       NaN   
3            501201         501283  2016-01-05 00:00:00   -83.0  Returned   
4            501301         501400  2014-12-24 00:00:00   100.0       NaN   

                       clinic tag_year     action action_date  action_year  \
0  Southern Animal Foundation     2015  Delivered  2014-12-12       2014.0   
1     Banfield 1023 (Elmwood)     2015  Delivered  2014-12-13       2014.0   
2          Animal Care Center     2015  Delivered  2014-12-29       2014.0   
3          Animal Care Center     2015   Returned  2016-01-05       2016.0   
4       Allstar Animal Clinic     2015  Delivered  2014-12-24       2014.0   

   action_month  
0          12.0  
1          12.0  
2          12.0  
3           1.0  
4          12.0

In [92]:
tag_data.loc[tag_data['delivered'].isna(), 'action_date'] = 'Unknown'
tag_data.loc[tag_data['delivered'].isna(), 'action_year'] = 'Unknown'
tag_data.loc[tag_data['delivered'].isna(), 'action_month'] = 'Unknown'

In [93]:
tag_data.head()

beginning_tag_num ending_tag_num            delivered  volume  returned  \
0            500001         501000  2014-12-12 00:00:00  1000.0       NaN   
1            501001         501200  2014-12-13 00:00:00   200.0       NaN   
2            501201         501300  2014-12-29 00:00:00   100.0       NaN   
3            501201         501283  2016-01-05 00:00:00   -83.0  Returned   
4            501301         501400  2014-12-24 00:00:00   100.0       NaN   

                       clinic tag_year     action          action_date  \
0  Southern Animal Foundation     2015  Delivered  2014-12-12 00:00:00   
1     Banfield 1023 (Elmwood)     2015  Delivered  2014-12-13 00:00:00   
2          Animal Care Center     2015  Delivered  2014-12-29 00:00:00   
3          Animal Care Center     2015   Returned  2016-01-05 00:00:00   
4       Allstar Animal Clinic     2015  Delivered  2014-12-24 00:00:00   

  action_year action_month  
0      2014.0         12.0  
1      2014.0         12.0  
2      2014.0         12.0  
3      2016.0          1.0  
4      2014.0         12.0

In [94]:
## Read in location data to match with tag data
location = pd.read_excel('../Data/tags_delivered.xlsx', sheet_name='Contact Info')
location.head()

C:\Users\rvick\Anaconda3\envs\R_Python\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: Your version of xlrd is 1.2.0. In xlrd >= 2.0, only the xls format is supported. As a result, the openpyxl engine will be used if it is installed and the engine argument is not specified. Install openpyxl instead.
  


Clinic Name                         Address 1         City  \
0  Acadiana West Animal Clinic  2600 Barataria Boulevard Suite E      Marrero   
1        Algiers Animal Clinic             3807 General DeGaulle      Algiers   
2       All Cats Mobile Clinic                               NaN          NaN   
3        Allstar Animal Clinic             4701 S Carrollton Ave  New Orleans   
4           Animal Care Center                     313 David Dr.     Metairie   

  State      ZIP  Phone Number Fax Number                  Contact  \
0    LA  70072.0     341-9510    328-0820                      NaN   
1    LA  70114.0      364-1556   361-3750                      NaN   
2   NaN      NaN  985-788-4322        NaN  Karen Miller-Becnel,DVM   
3    LA  70119.0     483-2224    483-6607         Ashley Ferrario    
4    LA  70003.0      733-6677   733-6738     animalcareVh@aol.com   

        Mailing Contact                     Email Address  \
0    Susan Olivier, DVM            acadianawest@yahoo.com   
1  Andrew E Gutter, DVM  dustyaac@gmail.com (Dusty Brown)   
2                   NaN                 catsDVM@yahoo.com   
3    Treyton Diggs, DVM    ashley@allstaranimalclinic.com   
4                   NaN    animalcarecentervh@hotmail.com   

                                            Comments  \
0                                                NaN   
1                                                NaN   
2  www.artforcatssake.org; www.catsmobileclinic.c...   
3             Afrika Bettis <afrikaakua@hotmail.com>   
4                                                NaN   

           Additional Contact info  Unnamed: 12  Unnamed: 13  Unnamed: 14  \
0                              NaN          NaN          NaN          NaN   
1                              NaN          NaN          NaN          NaN   
2                              NaN          NaN          NaN          NaN   
3  drdiggs@allstaranimalclinic.com          NaN          NaN          NaN   
4                              NaN          NaN          NaN          NaN   

   Unnamed: 15 Unnamed: 16  
0          NaN         NaN  
1          NaN         NaN  
2          NaN         NaN  
3          NaN         NaN  
4          NaN         NaN

In [95]:
#filter to name and zip only
location = location[['Clinic Name','Address 1', 'City', 'ZIP']]
location.columns = ['clinic','address','city', 'zip']
location.head()

clinic                           address         city  \
0  Acadiana West Animal Clinic  2600 Barataria Boulevard Suite E      Marrero   
1        Algiers Animal Clinic             3807 General DeGaulle      Algiers   
2       All Cats Mobile Clinic                               NaN          NaN   
3        Allstar Animal Clinic             4701 S Carrollton Ave  New Orleans   
4           Animal Care Center                     313 David Dr.     Metairie   

       zip  
0  70072.0  
1  70114.0  
2      NaN  
3  70119.0  
4  70003.0

In [96]:
tag_data['clinic_lower'] = tag_data['clinic'].str.lower()
location['clinic_lower'] = location['clinic'].str.lower()
tag_data['clinic_lower'] = tag_data['clinic_lower'].str.strip()
location['clinic_lower'] = location['clinic_lower'].str.strip()

In [97]:
merge_data = pd.merge(tag_data, location, how='left', on='clinic_lower')

In [98]:
merge_data.shape

(1525, 16)

In [99]:
merge_data[merge_data['clinic_y'].isna()]['clinic_x'].unique()

array(['Annual Rabies Vaccination Drive', 'Villalobos',
       'Banfield 1034 (West Bank)', 'W. Esplanade Vet Clinic',
       'VCA Airline Animal Hospital', 'Vetco @ Petco', 'Louisiana SPCA',
       'The SULA Foundation', 'Marrero Veterinary Clinic',
       'Laplace Veterinary', 'Unknown', 'Banfield - Remote',
       'Batt Veterinary Services', 'Canal Street Animal Clinic',
       'Geaux Nola Pets Mobile', 'Vet Pronto', nan,
       'Lakeside Animal Hospital', 'The Inner Pup',
       'Thrive Affordable Vet Care (Midcity)',
       'Thrive Affordable Vet Care (Vets)', 'The Street Dog Coalition',
       'Plaquemines'], dtype=object)

In [100]:
location[location['clinic_lower'].str.contains('banfield')]

clinic                          address      city  \
12   Banfield 1023 (Elmwood)  1000 S Clearview Prkwy Ste 1050   Harahan   
13  Banfield 1034 (Westbank)              1321 Manhattan Blvd    Harvey   
14  Banfield 1175 (Veterans)      6851 Veterans Memorial Blvd  Metairie   

        zip              clinic_lower  
12  70123.0   banfield 1023 (elmwood)  
13  70058.0  banfield 1034 (westbank)  
14  70003.0  banfield 1175 (veterans)

In [101]:
location[location['clinic_lower'].str.contains('esplanade')]

clinic                address      city  \
79  W. Esplanade Vet Clinic & Bird Hospital  3640 W. Esplanade Ave  Metairie   

        zip                             clinic_lower  
79  70002.0  w. esplanade vet clinic & bird hospital

In [102]:
location[location['clinic_lower'].str.contains('vca')]

clinic                address      city      zip  \
72           VCA Airline Animal      5105 Airline Drive  Metairie  70001.0   
73  VCA Causeway Animal Hospital  1315 N. Causeway Blvd  Metairie  70001.0   

                    clinic_lower  
72            vca airline animal  
73  vca causeway animal hospital

In [103]:
location[location['clinic_lower'].str.contains('vetco')]

clinic                address         city      zip  \
76  Vetco @ Petco Mobile Clinic  17485 Opportunity Ave  Baton Rouge  70817.0   
77       Vetco @ Petco National                    NaN          NaN      NaN   

                   clinic_lower  
76  vetco @ petco mobile clinic  
77       vetco @ petco national

In [104]:
location[location['clinic_lower'].str.contains('sula')]
#no address anyway

clinic address city  zip  \
68  The SULA Foundation (closed)     NaN  NaN  NaN   

                    clinic_lower  
68  the sula foundation (closed)

In [105]:
location[location['clinic_lower'].str.contains('marrero')]


clinic            address    city      zip  \
48  Marrero Vet. Clinic  4001 Lapalco Blvd  Harvey  70058.0   

           clinic_lower  
48  marrero vet. clinic

In [106]:
location[location['clinic_lower'].str.contains('laplace')]


clinic               address      city      zip  \
42  Laplace Veterinary Hospital  2921 New Highway 51   La Place  70068.0   

                   clinic_lower  
42  laplace veterinary hospital

In [107]:
location[location['clinic_lower'].str.contains('batt')]


Empty DataFrame
Columns: [clinic, address, city, zip, clinic_lower]
Index: []

In [108]:
location[location['clinic_lower'].str.contains('canal')]


clinic        address         city      zip  \
18  Canal Street Veterinary Hospital  4022 Canal St  New Orleans  70119.0   

                        clinic_lower  
18  canal street veterinary hospital

In [109]:
#Geaux Nola Pets Mobile
location[location['clinic_lower'].str.contains('geaux')]

clinic                address  \
31  Geaux Nola Pets Mobile Veterinary Services, LLC  2111 General Pershing   

           city      zip                                     clinic_lower  
31  New Orleans  70115.0  geaux nola pets mobile veterinary services, llc

In [110]:
#Geaux Nola Pets Mobile
location[location['clinic_lower'].str.contains('pronto')]
#no address anyway

clinic address city  zip clinic_lower
78  VetPronto     NaN  NaN  NaN    vetpronto

In [111]:
#Geaux Nola Pets Mobile
location[location['clinic_lower'].str.contains('lakeside')]

clinic                   address      city      zip clinic_lower
40  Lakeside  3838 Veterans Mem. Blvd.  Metairie  70002.0     lakeside

In [112]:
#Geaux Nola Pets Mobile
location[location['clinic_lower'].str.contains('inner')]

Empty DataFrame
Columns: [clinic, address, city, zip, clinic_lower]
Index: []

In [113]:
#Geaux Nola Pets Mobile
location[location['clinic_lower'].str.contains('thrive')]

clinic                       address      city  \
69  Thrive Affordable Vet Care                2110 S Sheppard   Houston   
70   Thrive Affordable Vet Care  3520 Veterans Memorial Blvd.  Metairie   

        zip                clinic_lower  
69  77019.0  thrive affordable vet care  
70  70002.0  thrive affordable vet care

In [114]:
location[location['clinic_lower'].str.contains('street')]

clinic           address         city      zip  \
18  Canal Street Veterinary Hospital     4022 Canal St  New Orleans  70119.0   
45     Magazine Street Animal Clinic  3458 Magazine St  New Orleans  70115.0   

                        clinic_lower  
18  canal street veterinary hospital  
45     magazine street animal clinic

In [115]:
location['clinic'] = location['clinic'].replace('Banfield 1034 (Westbank)', 'Banfield 1034 (West Bank)')
location['clinic'] = location['clinic'].replace('W. Esplanade Vet Clinic & Bird Hospital','W. Esplanade Vet Clinic')
location['clinic'] = location['clinic'].replace('VCA Airline Animal ','VCA Airline Animal Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Vetco @ Petco','Vetco @ Petco Mobile Clinic')
location['clinic'] = location['clinic'].replace('Marrero Vet. Clinic','Marrero Veterinary Clinic')
tag_data['clinic'] = tag_data['clinic'].replace('Laplace Veterinary','Laplace Veterinary Hospital')
tag_data['clinic'] = tag_data['clinic'].replace('Canal Street Animal Clinic','Canal Street Veterinary Hospital')
location['clinic'] = location['clinic'].replace('Geaux Nola Pets Mobile Veterinary Services, LLC','Geaux Nola Pets Mobile')
location['clinic'] = location['clinic'].replace('Lakeside','Lakeside Animal Hospital')
location['clinic'] = location['clinic'].replace('Thrive Affordable Vet Care','Lakeside Animal Hospital')
location.loc[location['address'] == '3520 Veterans Memorial Blvd.', 'clinic' ] = 'Thrive Affordable Vet Care (Vets)'

In [116]:
tag_data['clinic_lower'] = tag_data['clinic'].str.lower()
location['clinic_lower'] = location['clinic'].str.lower()
tag_data['clinic_lower'] = tag_data['clinic_lower'].str.strip()
location['clinic_lower'] = location['clinic_lower'].str.strip()

In [117]:
merge_data = pd.merge(tag_data, location, how='left', on='clinic_lower')

In [118]:
merge_data.head()

beginning_tag_num ending_tag_num            delivered  volume  returned  \
0            500001         501000  2014-12-12 00:00:00  1000.0       NaN   
1            501001         501200  2014-12-13 00:00:00   200.0       NaN   
2            501201         501300  2014-12-29 00:00:00   100.0       NaN   
3            501201         501283  2016-01-05 00:00:00   -83.0  Returned   
4            501301         501400  2014-12-24 00:00:00   100.0       NaN   

                     clinic_x tag_year     action          action_date  \
0  Southern Animal Foundation     2015  Delivered  2014-12-12 00:00:00   
1     Banfield 1023 (Elmwood)     2015  Delivered  2014-12-13 00:00:00   
2          Animal Care Center     2015  Delivered  2014-12-29 00:00:00   
3          Animal Care Center     2015   Returned  2016-01-05 00:00:00   
4       Allstar Animal Clinic     2015  Delivered  2014-12-24 00:00:00   

  action_year action_month                clinic_lower  \
0      2014.0         12.0  southern animal foundation   
1      2014.0         12.0     banfield 1023 (elmwood)   
2      2014.0         12.0          animal care center   
3      2016.0          1.0          animal care center   
4      2014.0         12.0       allstar animal clinic   

                     clinic_y                          address         city  \
0  Southern Animal Foundation             1823 Magazine Street  New Orleans   
1     Banfield 1023 (Elmwood)  1000 S Clearview Prkwy Ste 1050      Harahan   
2          Animal Care Center                    313 David Dr.     Metairie   
3          Animal Care Center                    313 David Dr.     Metairie   
4       Allstar Animal Clinic            4701 S Carrollton Ave  New Orleans   

       zip  
0  70130.0  
1  70123.0  
2  70003.0  
3  70003.0  
4  70119.0

In [119]:
merge_data[merge_data['clinic_y'].isna()]['clinic_x'].unique()

array(['Annual Rabies Vaccination Drive', 'Villalobos', 'Louisiana SPCA',
       'The SULA Foundation', 'Unknown', 'Banfield - Remote',
       'Batt Veterinary Services', 'Vet Pronto', nan, 'The Inner Pup',
       'Thrive Affordable Vet Care (Midcity)', 'The Street Dog Coalition',
       'Plaquemines'], dtype=object)

In [120]:
location[location['clinic_lower'].str.contains('plaquemines')]

Empty DataFrame
Columns: [clinic, address, city, zip, clinic_lower]
Index: []

In [121]:
#view null clinics
merge_data[merge_data['clinic_x'].isna()]

beginning_tag_num ending_tag_num            delivered  volume  returned  \
740             727213         727225                  NaN   -12.0  Returned   
999             825190         825200  2019-04-01 00:00:00   -11.0  Returned   
1000            813XXX         813XXX  2019-01-04 00:00:00    -2.0  Returned   
1244            9433XX         9433XX  2020-02-18 00:00:00   -12.0  Returned   
1245            9564XX         9564XX  2020-01-30 00:00:00   -12.0  Returned   

     clinic_x tag_year    action          action_date action_year  \
740       NaN     2017  Returned              Unknown     Unknown   
999       NaN     2018  Returned  2019-04-01 00:00:00      2019.0   
1000      NaN     2018  Returned  2019-01-04 00:00:00      2019.0   
1244      NaN     2019  Returned  2020-02-18 00:00:00      2020.0   
1245      NaN     2019  Returned  2020-01-30 00:00:00      2020.0   

     action_month clinic_lower clinic_y address city  zip  
740       Unknown          NaN      NaN     NaN  NaN  NaN  
999           4.0          NaN      NaN     NaN  NaN  NaN  
1000          1.0          NaN      NaN     NaN  NaN  NaN  
1244          2.0          NaN      NaN     NaN  NaN  NaN  
1245          1.0          NaN      NaN     NaN  NaN  NaN

In [122]:
#remove unknown clinic records
merge_data = merge_data[merge_data['clinic_x'].isna()==False]

In [123]:
merge_data.shape

(1520, 16)

In [124]:
#drop unneeded columns and columnsthat were used to derive other data
merge_data = merge_data.drop(columns=['beginning_tag_num', 'ending_tag_num', 'delivered', 'returned', 'clinic_lower', 'clinic_y'])

In [125]:
merge_data = merge_data.rename(columns={'clinic_x':'clinic'})

In [126]:
merge_data

volume                      clinic tag_year     action  \
0     1000.0  Southern Animal Foundation     2015  Delivered   
1      200.0     Banfield 1023 (Elmwood)     2015  Delivered   
2      100.0          Animal Care Center     2015  Delivered   
3      -83.0          Animal Care Center     2015   Returned   
4      100.0       Allstar Animal Clinic     2015  Delivered   
...      ...                         ...      ...        ...   
1520    -1.0       Algiers Animal Clinic     2020   Returned   
1521   -17.0       Algiers Animal Clinic     2020   Returned   
1522   -39.0       Algiers Animal Clinic     2020   Returned   
1523   -15.0       Algiers Animal Clinic     2020   Returned   
1524   -51.0          NOLA Animal Clinic     2020   Returned   

              action_date action_year action_month  \
0     2014-12-12 00:00:00      2014.0         12.0   
1     2014-12-13 00:00:00      2014.0         12.0   
2     2014-12-29 00:00:00      2014.0         12.0   
3     2016-01-05 00:00:00      2016.0          1.0   
4     2014-12-24 00:00:00      2014.0         12.0   
...                   ...         ...          ...   
1520  2021-01-04 00:00:00      2021.0          1.0   
1521  2021-01-04 00:00:00      2021.0          1.0   
1522  2021-01-04 00:00:00      2021.0          1.0   
1523  2021-01-04 00:00:00      2021.0          1.0   
1524  2021-01-04 00:00:00      2021.0          1.0   

                              address         city      zip  
0                1823 Magazine Street  New Orleans  70130.0  
1     1000 S Clearview Prkwy Ste 1050      Harahan  70123.0  
2                       313 David Dr.     Metairie  70003.0  
3                       313 David Dr.     Metairie  70003.0  
4               4701 S Carrollton Ave  New Orleans  70119.0  
...                               ...          ...      ...  
1520            3807 General DeGaulle      Algiers  70114.0  
1521            3807 General DeGaulle      Algiers  70114.0  
1522            3807 General DeGaulle      Algiers  70114.0  
1523            3807 General DeGaulle      Algiers  70114.0  
1524                  4554 Downman Rd  New Orleans  70126.0  

[1520 rows x 10 columns]

In [127]:
#save the merged dataset to use in graph analysis
merge_data.to_csv('../Data/analysis_dataset.csv', index=False)

In [128]:
merge_data.shape

(1520, 10)

In [129]:
merge_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1520 entries, 0 to 1524
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   volume        1520 non-null   float64
 1   clinic        1520 non-null   object 
 2   tag_year      1520 non-null   object 
 3   action        1520 non-null   object 
 4   action_date   1519 non-null   object 
 5   action_year   1519 non-null   object 
 6   action_month  1519 non-null   object 
 7   address       1193 non-null   object 
 8   city          1165 non-null   object 
 9   zip           1139 non-null   float64
dtypes: float64(2), object(8)
memory usage: 130.6+ KB
